### DISCUSSION TITLE
Flicker on Mouse Move when hosting in WinForms

### DISCUSSION THREAD

**Discussion by fitdev** [AvaloniaUI/Avalonia#14289] [Q&A] [answered]
I want to have Avalonia "Islands" in my WinForms application where I would like to display richer content possible with Avalonia.

To that end I have used `AvaloniaControlHost` (basically what is in https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32.Interoperability/WinForms/WinFormsAvaloniaControlHost.cs and `Avalonia.Win32.Interoperability` package):

```cs
public class AvaloniaControlHost : System.Windows.Forms.Control {

  readonly EmbeddableControlRoot _root = new();

  readonly IntPtr _windowHandle;

  public AvaloniaControlHost() {
  
    SetStyle(ControlStyles.AllPaintingInWmPaint | ControlStyles.UserPaint | ControlStyles.Opaque | ControlStyles.OptimizedDoubleBuffer, true); //<-- Changing flags here does not seem to affect flicker at all
    
    _windowHandle = _root.TryGetPlatformHandle()?.Handle ?? IntPtr.Zero;

    SetParent(_windowHandle, Handle);

    _root.Prepare();
    if (_root.IsFocused) {
      _root.FocusManager?.ClearFocus();
    }
    _root.GotFocus += (_, _) => SetFocus(_windowHandle);

    FixPosition();
  }

  public AvaloniaControlHost(Avalonia.Controls.Control? content) : this() {
    Content = content;
  }

  /// <summary>
  /// Set this to whatever Avalonia-based needs to be hosted/displayed in place of this control.
  /// </summary>
  public Avalonia.Controls.Control? Content {
    get => (Avalonia.Controls.Control?)_root.Content;
    set {
      _root.Content = value;
      _root.StartRendering();
    }
  }

  protected override void DestroyHandle() {
    _root.StopRendering();
    base.DestroyHandle();
  }

  protected override void Dispose(bool disposing) {
    if (disposing) {
      _root.GotFocus -= (_, _) => SetFocus(_windowHandle);
      _root.Dispose();
    }
    base.Dispose(disposing);
  }

  protected override void OnGotFocus(EventArgs e) => SetFocus(_windowHandle);

  protected override void OnResize(EventArgs e) {
    FixPosition();
    base.OnResize(e);
  }

  protected override void OnPaint(PaintEventArgs e) {
    // Avalonia renderer performs all painting
  }

  void FixPosition() {
    if (Width > 0 && Height > 0) MoveWindow(_windowHandle, 0, 0, Width, Height, true);
  }

  [DllImport("user32.dll")]
  static extern IntPtr SetParent(IntPtr hWndChild, IntPtr hWndNewParent);

  [DllImport("user32.dll")]
  static extern bool SetFocus(IntPtr hWnd);

  [DllImport("user32.dll", SetLastError = true)]
  static extern bool MoveWindow(IntPtr hWnd, int X, int Y, int nWidth, int nHeight, bool bRepaint);

}

public class AvaloniaControlHost<TControl> : AvaloniaControlHost where TControl : Avalonia.Controls.Control, new() {

  public AvaloniaControlHost() : base(new TControl()) { }
  public AvaloniaControlHost(TControl control) : base(control) { }

}

```
BTW, I did not use `Avalonia.Win32.Interoperability` package for 2 reasons:
1. Back when I started to need this about a year ago it still used reflection to obtain the underlying handle, which was not acceptable for us.
2. There is basically just one source file - https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32.Interoperability/WinForms/WinFormsAvaloniaControlHost.cs - and to refernce a whole package just for this seems too much.
So the above code is given to aid the understanding, and is basically identical to using `Avalonia.Win32.Interoperability` package.

Then I use the following bootstrap code (run on the UI thread before first form is displayed, adapted from https://github.com/AvaloniaUI/Avalonia/blob/master/samples/interop/WindowsInteropTest/Program.cs):

```cs
AppBuilder
	.Configure<MyApp>()
	.UsePlatformDetect()
  .With(new Win32PlatformOptions {
    //NOTE: These fix some flicker but cause ugly artifacts with rounded corners and not so smooth appearance of popups like tooltips
    CompositionMode = [Win32CompositionMode.RedirectionSurface],
    ShouldRenderOnUIThread = true // necessary for WPF
  })
  .SetupWithoutStarting();
```

Then to actually use Avalonia controls, I do, say in my Form's constructor:

```cs

public class MyForm : Form {

  AvaloniaControlHost Host;

  public MyForm() {
    InitializeComponent();

    Host = new AvaloniaControlHost<MyAvaloniaControlIWantToUse> {
      Dock = DockStyle.Fill
    };
    SuspendLayout();
    Controls.Add(Host);
    ResumeLayout(false);

  }

}
```

This all works well, except that there is a terrible flicker whenever I move the mouse over the Avalonia-rendered part.

I noticed that if I set `ShouldRenderOnUIThread = true` in `Win32PlatformOptions`, this does reduce the flicker - now the flicker seems to affect _only individual Avalonia elements_ (where the mouse is moving over) instead of the whole control, but now there are issues with rounded corners on popups like tooltips - they have ugly white backgrounds.

So, what am I doing wrong? How do I eliminate flickering with mouse moves over Avalonia-rendered part?

Thanks.

**Comment by thevortexcloud** [comment]
There is an official way to do winforms interop with Avalonia. I believe you just need this package: https://www.nuget.org/packages/Avalonia.Win32.Interoperability/

There is an example here on how to use it: https://github.com/AvaloniaUI/Avalonia/tree/master/samples/interop/WindowsInteropTest

There used to also be official WPF support but it was removed.

**Reply by fitdev** [reply]
Thank you for your comment. I know and that is what I basically used. However my question is specifically about flicker, not how to do WinForms interop.

**Reply by fitdev** [reply]
BTW, I did not use `Avalonia.Win32.Interoperability` package for 2 reasons:
1. Back when I started to need this about a year ago it still used reflection to obtain the underlying handle, which was not acceptable for us.
2. There is basically just one source file - https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32.Interoperability/WinForms/WinFormsAvaloniaControlHost.cs - and to refernce a whole package just for this seems too much.
The above code is the same as what is in the https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32.Interoperability/WinForms/WinFormsAvaloniaControlHost.cs via `Avalonia.Win32.Interoperability` package.

So the question is not about how to embed Avalonia in WinForms. The question is how to avoid flicker during mouse move events.

**Comment by fitdev** [comment, marked as answer]
For those interested, there are 2 possible solutions:

- Either set `CompositionMode = [Win32CompositionMode.RedirectionSurface]` in `AppBuilder` configuration when configuring `Win32PlatformOptions`. But this has the unwanted effect of not being able to have nicely rounded popups and window or windows with transparency / shadows.
- OR: As was suggested in https://stackoverflow.com/questions/22233806/mouse-cursor-flickers-over-selected-text-how-to-prevent-this disable processing of `WM_SETCURSOR` message by creating a derived / your own version of `Avalonia.Win32.Interoperability.WinFormsAvaloniaControlHost` (https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32.Interoperability/WinForms/WinFormsAvaloniaControlHost.cs) and override `WndProc(ref System.Windows.Forms.Message m)` to not process `WM_SETCURSOR` messages. The benefit of this solution is that it works with `Win32CompositionMode.WinUIComposition` mode allowing for nice transparency-supporting windows.
